### ISSUE TITLE
Cloud_RoleName is incorrect

### ISSUE THREAD

**Issue by Vivihung** [Microsoft/ApplicationInsights-dotnet-server#648]
Cloud_RoleName is incorrect in app insight.

Background:
We have two deployment slots in Azure.
Production: <product-name>-<region>
Staging: <product-name>-<region>-staging

We deploy to staging slot, verified the service, swap slots and stop the staging one.
However, in app insight, the production slot instances are shown with staging role name after we swap.
It seems a bug to me.

**Comment by SergeyKanzhelev** [comment]
@iusafaro is it expected?

**Comment by iusafaro** [comment]
@Vivihung, are you talking about Azure Web Apps? ApplicationInsights is getting role name for Azure web apps from WEBSITE_HOSTNAME environment variable. Can you please check what is set there for you in production instance by navigating to Kudu/Environment?

**Comment by Vivihung** [comment]
@iusafaro We are using Azure app service. And I believe behind the scene is Azure Web App.
I've confirmed in Kudu/Environment that all the WEBSITE_HOSTNAME is configured correctly.

Do you cache the site name in memory?
If I swap deployment slots, how does your system update WEBSITE_HOSTNAME?

**Comment by cijothomas** [comment]
Yes, the value is read once and cached.
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs#L47

As of now, i don't think there i a way to know deployment slots have swapped. @iusafaro do you know if can we listen on some event to know swap happened?

**Comment by jawn** [comment]
So the workaround for now is to restart both swapped instances?

**Comment by Vivihung** [comment]
Restart instances after swap will bring our service offline, which is not acceptable in our case :(

**Comment by SergeyKanzhelev** [comment]
@jawn yes, easiest approach is to restart the instances
@Vivihung sure, one approach would be to implement a custom telemetry initializer that will populate role name properly. Docs are [here](https://docs.microsoft.com/azure/application-insights/app-insights-api-filtering-sampling#add-properties-itelemetryinitializer)

Basically you need something like this:

``` csharp
public class RoleNameWithRefresh : ITelemetryInitializer
{
    string roleName = "";
    private RefreshRoleName()
    {
        //check some timer and refresh if required
        if (enough time passed)
          this.roleName = ReadRoleNameFromEnvironmentVariables();
    }
    public void Initialize(ITelemetry telemetry)
    {
        this.RefreshRoleName();
        telemetry.Cloud.RoleInstance = this.roleName;
    }
}
```

Here is the original code [for reference](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/91016d62f3181e10d4cf589ef8fd64dadb6b54a2/Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs)

**Comment by SergeyKanzhelev** [comment]
@davidebbo is there any sort of notification that we can subscribe on to re-read role name after swap of slots?

**Comment by davidebbo** [comment]
@SergeyKanzhelev I don't believe so. @ruslany, is there anything that they can look for to detect when a swap happens? I guess it depends on whether the slot gets restarted as part of the swap, which depends on the presence of slot settings.

**Comment by SergeyKanzhelev** [comment]
@davidebbo @ruslany if we need to implement periodic check on whether swap just happened - is reading of environment variable is the fastest approach? Any other indications SDK can watch for that can be more performant?

Also - is Kudu knows about the swap? If it is - maybe we can communicate back from Kudu to the worker process.

**Comment by davidebbo** [comment]
The problem is that I don't think the `WEBSITE_HOSTNAME` env variable is even guaranteed to be updated when the swap complete, if there are no slot settings. @ruslany, can you confirm?

**Comment by SergeyKanzhelev** [comment]
@davidebbo @Vivihung mentioned above that `WEBSITE_HOSTNAME` seems to be handled correctly. But maybe it's what Kudu shows. Anyway, @ruslany what will be the recommendation here?

**Comment by brettsam** [comment]
FYI -- we just worked around this in Functions with https://github.com/Azure/azure-webjobs-sdk/pull/1382

**Comment by soorena776** [comment]
@brettsam how can I use this fix? I don't see any releases after this.

**Comment by soorena776** [comment]
I'm trying to set cloud_RoleName and cloud_RoleInstance for an azure webjob after integrating with app insights:
[https://github.com/Azure/azure-webjobs-sdk/wiki/Application-Insights-Integration](url)

I do see -staging postfix in insights' logs. I tried creating a custom ITelemetryInitializer but it didn't work. As a workaround, I added the following code in WebJobs' Program.cs
`Environment.SetEnvironmentVariable("WEBSITE_HOSTNAME", ConfigurationManager.AppSettings["EnvironmentName"]);`

**Comment by s-krawczyk** [comment]
+1

**Comment by tomaszzmuda** [comment]
Still doesn't work.

**Comment by AndrzejWroGM** [comment]
+1

**Comment by AdrianStrugala** [comment]
+1

**Comment by dpaquette** [comment]
Any plans of fixing this issue?

**Comment by cijothomas** [comment]
At this point - AI Telemetry initializer is doing the best it can. If the ENV variable it relies on is not updated underneath - then not much can be done on this initializer..

**Comment by cijothomas** [comment]
I will check with app service folks to see if there is anything better we can do here.

**Comment by dpaquette** [comment]
I just double checked the ENV variables in Kudu. In my case, the `WEBSITE_HOSTNAME` ENV variable is incorrect (it still has the deployment slot URL in it), but the `WEBSITE_SITE_NAME` variable appears correct.

**Comment by MarkDordoy** [comment]
I am also seeing this issue. My slot is stopped after swap, but when running the following app insights query: 
```requests
| order by timestamp desc
| where resultCode startswith "2" 
| project timestamp, resultCode ,cloud_RoleName,url, user_Id, name ```

some of the requests have the cloud_RoleName showing as my staging slot, which i've manually gone and checked and its stopped. 

When will this be fixed?

**Comment by razzemans** [comment]
We are running into this issue as well. Would very much like a fix.

**Comment by cijothomas** [comment]
The proposed fix https://github.com/microsoft/ApplicationInsights-dotnet-server/issues/1207 is scheduled to arrive for next stable, most likely in the beta2.
